### PR TITLE
 Hoisting logic needs to consider dependency on prior unhoistable variable

### LINKED ISSUE #9581: Non-det bug with prange code 2
The issue reported in #9490 has only been partially resolved. It appears that the problem persists, and we may need to reopen the issue.

I've added further details to the original issue for your reference.

### PR DESCRIPTION
**Pull request by sklam**
Fix #9581, #9490.

### CODE REVIEW

File: numba/parfors/parfor_lowering.py [hunk 1]
```diff
@@ -888,7 +888,10 @@ def _hoist_internal(inst, dep_on_param, call_table, hoisted, not_hoisted,
 
     uses = set()
     visit_vars_inner(inst.value, find_vars, uses)
+    unhoistable = {assgn.target.name for assgn, _reason in not_hoisted}
+    use_unhoist = uses & unhoistable
     diff = uses.difference(dep_on_param)
+    diff |= use_unhoist
```
Comment (sklam) [inline numba/parfors/parfor_lowering.py:894]: @DrTodd13, I need your expertise here. I only did the smallest change I can think of. 

The origin is that a `static_getitem(array, index)` is referring to an `array` that is previously placed into `not_hoisted`. The source code looks like:

```python
for .. in loop:
    array = solve(arg)
    res = array[1]     # the static getitem
    if compute(res):
          break
use(array)
```

Pre-patch logic will cause that `static_getitem` to be hoisted out of the loop. I changed it to look into the list of `not_hoisted` and ban hoisting on `uses` that refer to it.

File: numba/parfors/parfor_lowering.py [hunk 2]
```diff
@@ -888,7 +888,10 @@ def _hoist_internal(inst, dep_on_param, call_table, hoisted, not_hoisted,
 
     uses = set()
     visit_vars_inner(inst.value, find_vars, uses)
+    unhoistable = {assgn.target.name for assgn, _reason in not_hoisted}
```
Comment (DrTodd13) [inline numba/parfors/parfor_lowering.py:891]:
```suggestion
    unhoistable = {assgn.target.name for assgn, _ in not_hoisted}
```

**Review by DrTodd13 — CHANGES_REQUESTED** [review]
Looks good in general.  Just my one suggestion.

**Comment by sklam** [comment]
@DrTodd13, i just realize you have a PR that sounds similar. Will https://github.com/numba/numba/pull/9397 fix this as well?

**Comment by DrTodd13** [comment]
> @DrTodd13, i just realize you have a PR that sounds similar. Will #9397 fix this as well?

My instinct is to say that if the dependency is two lines right after each other then that other PR wouldn't help but it is worth trying it.  I'd also suggest maybe rebasing this PR against mine and doing them together rather than separately.

**Comment by sklam** [comment]
I confirmed that #9397 do not fix the issues https://github.com/numba/numba/issues/9581, https://github.com/numba/numba/issues/9490.

**Review by gmarkall — APPROVED** [review]
Without the fix in `parfor_lowering`, the modified test case fails with:

```
test_issue9490_non_det_ssa_problem (numba.tests.test_parfors.TestParfors.test_issue9490_non_det_ssa_problem) ... FAIL

======================================================================
FAIL: test_issue9490_non_det_ssa_problem (numba.tests.test_parfors.TestParfors.test_issue9490_non_det_ssa_problem)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/gmarkall/numbadev/numba/numba/tests/test_parfors.py", line 2397, in test_issue9490_non_det_ssa_problem
    subp.check_output(cmd, env=envs,
  File "/home/gmarkall/miniforge3/envs/numbadev/lib/python3.11/subprocess.py", line 466, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/gmarkall/miniforge3/envs/numbadev/lib/python3.11/subprocess.py", line 571, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['/home/gmarkall/miniforge3/envs/numbadev/bin/python', '-m', 'numba.tests.parfor_iss9490_usecase']' died with <Signals.SIGSEGV: 11>.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/gmarkall/numbadev/numba/numba/tests/test_parfors.py", line 2402, in test_issue9490_non_det_ssa_problem
    self.fail(msg=msg)
AssertionError: subprocess failed with output:
Fatal Python error: Segmentation fault

Current thread 0x00007178c7191740 (most recent call first):
  File "/home/gmarkall/numbadev/numba/numba/tests/parfor_iss9490_usecase.py", line 69 in check
  File "/home/gmarkall/numbadev/numba/numba/tests/parfor_iss9490_usecase.py", line 78 in <module>
  File "<frozen runpy>", line 88 in _run_code
  File "<frozen runpy>", line 198 in _run_module_as_main
```

and the reproducer from #9581 also segfaults.

With the change in this PR, the test case passes and the reproducer from #9581 executes to completion.

I don't have enough knowledge of parfors to have an opinion on the fix, but from a "works for me" perspective, this PR appears to do the job.

**Pull request merged**